test/cpp/end2end/generic_end2end_test.cc

```diff
@@ -330,6 +381,29 @@ TEST_F(GenericEnd2endTest, Deadline) {
                        gpr_time_from_seconds(10, GPR_TIMESPAN)));
 }
 
+TEST_F(GenericEnd2endTest, ShortDeadline) {
+  ResetStub();
+
+  ClientContext cli_ctx;
+  EchoRequest request;
+  EchoResponse response;
+
+  shutting_down_ = false;
+  std::thread driver([this] { DriveCompletionQueue(); });
+
+  request.set_message("");
+  cli_ctx.set_deadline(gpr_time_add(gpr_now(GPR_CLOCK_MONOTONIC),
+                                    gpr_time_from_micros(500, GPR_TIMESPAN)));
+  Status s = stub_->Echo(&cli_ctx, request, &response);
+  EXPECT_FALSE(s.ok());
+  {
+    std::lock_guard<std::mutex> lock(shutting_down_mu_);
+    shutting_down_ = true;
+  }
+  TearDown();
```

Comment:
I wouldn't explicitly call TearDown like this since this is the thing that gets automatically called at the end of a gtest. I think you'll get double Shutdown calls then based on what you have right now. I realize that your intention is to shut down the server and completion queue. As such, I would suggest refactoring TearDown so that you can explicitly shutdown the things that you want to shut down here or in a separate function and then have TearDown just shutdown the things that aren't shutdown yet.

Reply:
Added a shut_down_ member and using that as a condition in ShutDownServerAndCQs(). This method is called from both TearDown and the ShortDeadlineTest.